experimental/lib/Transforms/Speculation/PlacementFinder.cpp

```diff
@@ -0,0 +1,359 @@
+  funcOp->walk([&](Operation *op) {
+    for (Value operand : op->getOperands()) {
+      BBEndpoints endpoints;
+      if (isBackedge(operand, op, &endpoints) or
+          endpoints.srcBB != endpoints.dstBB) {
+        OpPlacement arc = {operand, op};
+        endpointArcs[endpoints].push_back(arc);
+      }
+    }
+  });
+
+  // Join all predecessors of a BB
+  BBPredecessorMap predecessors;
+  for (const auto &[endpoints, arcs] : endpointArcs) {
+    BlockPredecessor predecessor;
+    predecessor.srcBB = endpoints.srcBB;
+    predecessor.dstBB = endpoints.dstBB;
+    predecessor.arcs = arcs;
+    predecessors[endpoints.dstBB].push_back(predecessor);
+  }
+
+  return predecessors;
+}
+
+// DFS traversal to mark all operations that lead to Commit units
+static void markSpeculativePaths(Operation *currOp,
+                                 SpeculationPlacements &placements,
+                                 PlacementList &markedPaths) {
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      OpPlacement arc = {res, succOp};
+      if (not markedPaths.count(arc)) {
+        markedPaths.insert(arc);
+        // Stop traversal if a commit is reached
+        if (not placements.containsCommit(res, succOp))
+          markSpeculativePaths(succOp, placements, markedPaths);
+      }
+    }
+  }
+}
+
+void PlacementFinder::findCommitsBetweenBBs() {
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  auto funcOp = specPos.dstOp->getParentOfType<handshake::FuncOp>();
+  assert(funcOp && "op should have parent function");
+
+  // Whenever a BB has two speculative inputs, commit units are needed to avoid
+  // tokens going out-of-order. First, the block predecessor arcs are found
+  BBPredecessorMap predecessors = getBlockPredecessors(funcOp);
+  PlacementList markedPaths;
+  markSpeculativePaths(specPos.dstOp, placements, markedPaths);
+  // Iterate all BBs to check if commits are needed
+  for (const auto &[bb, blockPredecessors] : predecessors) {
+    // Count number of speculative inputs to the BB
+    unsigned countSpecInputs = 0;
+    for (BlockPredecessor pred : blockPredecessors) {
+      if (llvm::any_of(pred.arcs,
+                       [&](OpPlacement p) { return markedPaths.count(p); }))
+        countSpecInputs++;
+    }
+
+    if (countSpecInputs > 1) {
+      // Potential ordering issue, add commits
+      for (BlockPredecessor pred : blockPredecessors) {
+        for (OpPlacement arc : pred.arcs) {
+          placements.addCommit(arc.srcOpResult, arc.dstOp);
+          // Here, synchronizer operations will be needed in the future
+        }
+      }
+    }
+  }
+
+  // Now that new commits have been added, some of the already placed commits
+  // might be unreachable. Hence, the path to commits is marked again and
+  // unreachable commits are removed
+  markedPaths.clear();
+  markSpeculativePaths(specPos.dstOp, placements, markedPaths);
+
+  // Remove commits that cannot be reached
```

Comment:
Haoran has a lot of discussion of doing this properly in the thesis, but this approach seems pretty naive: is it the same as what he has written?

Reply:
For the record, Emmet and I discussed this in a meeting and decided to address optimizations in a future pass.